### PR TITLE
"error jump" instead of STOP instraction in case of exception

### PR DESCRIPTION
**Pull request by LianaHus**

### CODE REVIEW

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -207,6 +210,10 @@ Json::Value Assembly::streamAsmJson(ostream& _out, StringMap const& _sourceCodes
 				createJsonValue("PUSH tag", i.getLocation().start, i.getLocation().end, m_strings.at((h256)i.data())));
 			break;
 		case PushTag:
+			if (i.data() == 0)
+				collection.append(
+					createJsonValue("PUSH [ErrorTag]", i.getLocation().start, i.getLocation().end, ""));
+
 			collection.append(
```
Comment (chriseth) [inline libevmasm/Assembly.cpp]: Shouldn't this be inside `else`?

File: libevmasm/Assembly.h [hunk 2]
```diff
@@ -67,6 +67,8 @@ class Assembly
 	AssemblyItem appendJumpI() { auto ret = append(newPushTag()); append(Instruction::JUMPI); return ret; }
 	AssemblyItem appendJump(AssemblyItem const& _tag) { auto ret = append(_tag.pushTag()); append(Instruction::JUMP); return ret; }
 	AssemblyItem appendJumpI(AssemblyItem const& _tag) { auto ret = append(_tag.pushTag()); append(Instruction::JUMPI); return ret; }
+	AssemblyItem errorTag() { return AssemblyItem(PushTag,  0); }
```
Comment (chriseth) [inline libevmasm/Assembly.h]: `,  0` (two spaces)

File: libsolidity/CompilerContext.h [hunk 3]
```diff
@@ -94,6 +94,8 @@ class CompilerContext
 	eth::AssemblyItem appendJumpToNew() { return m_asm.appendJump().tag(); }
 	/// Appends a JUMP to a tag already on the stack
 	CompilerContext&  appendJump(eth::AssemblyItem::JumpType _jumpType = eth::AssemblyItem::JumpType::Ordinary);
+	/// Appends a JUMP to an "ErrorTag"
```
Comment (chriseth) [inline libsolidity/CompilerContext.h]: It only returns the tag.

File: libsolidity/ExpressionCompiler.cpp [hunk 4]
```diff
@@ -1103,8 +1103,10 @@ void ExpressionCompiler::appendExternalFunctionCall(
 		m_context << eth::Instruction::CALLCODE;
 	else
 		m_context << eth::Instruction::CALL;
-	auto tag = m_context.appendConditionalJump();
-	m_context << eth::Instruction::STOP << tag;	// STOP if CALL leaves 0.
+
+	m_context << eth::Instruction::ISZERO;
+	auto tag = m_context.appendConditionalJumpTo(m_context.errorTag());// if CALL leaves 0.
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: Suggestion for comment: Propagate error condition (if CALL pushes 0 on stack).

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 5]
```diff
@@ -1788,6 +1788,7 @@ BOOST_AUTO_TEST_CASE(contracts_as_addresses)
 		}
 	)";
 	compileAndRun(sourceCode, 20);
+	auto res = callContractFunction("getBalance()");
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Remove?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 6]
```diff
@@ -4027,6 +4028,29 @@ BOOST_AUTO_TEST_CASE(overwriting_inheritance)
...
+				test = true;
+				return true;
+			}
+		}
+	)";
+	compileAndRun(sourceCode, 0, "A");
+	BOOST_CHECK(callContractFunction("test()") == encodeArgs(false));
+	BOOST_CHECK(callContractFunction("testIt()") == encodeArgs());
+	BOOST_CHECK(callContractFunction("test()") == encodeArgs(false));
+}
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Can you add a test where a non-existing function is called (via `address.call`)?

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 5 above)
Comment (LianaHus) [inline test/libsolidity/SolidityEndToEndTest.cpp]: this is for debugging should be removed later

File: libsolidity/CompilerContext.h [hunk 7]
```diff
@@ -94,6 +94,8 @@ class CompilerContext
 	eth::AssemblyItem appendJumpToNew() { return m_asm.appendJump().tag(); }
 	/// Appends a JUMP to a tag already on the stack
 	CompilerContext&  appendJump(eth::AssemblyItem::JumpType _jumpType = eth::AssemblyItem::JumpType::Ordinary);
+	/// Returns an "ErrorTag"
```
Comment (chriseth) [inline libsolidity/CompilerContext.h]: "the" error tag?

File: libsolidity/ExpressionCompiler.cpp [hunk 8]
```diff
@@ -1035,7 +1035,7 @@ void ExpressionCompiler::appendHighBitsCleanup(IntegerType const& _typeOnStack)
 void ExpressionCompiler::appendExternalFunctionCall(
 	FunctionType const& _functionType,
 	vector<ASTPointer<Expression const>> const& _arguments
-)
+	)
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: Should not be indented.

File: libsolidity/ExpressionCompiler.cpp [hunk 9]
```diff
@@ -1103,8 +1103,11 @@ void ExpressionCompiler::appendExternalFunctionCall(
 		m_context << eth::Instruction::CALLCODE;
 	else
 		m_context << eth::Instruction::CALL;
+
+	//Propagate error condition (if CALL pushes 0 on stack).
 	auto tag = m_context.appendConditionalJump();
-	m_context << eth::Instruction::STOP << tag;	// STOP if CALL leaves 0.
+	m_context << eth::Instruction::STOP << tag;	// STOP if CALL leaves 0.//	}
```
Comment (chriseth) [inline libsolidity/ExpressionCompiler.cpp]: This looks strange.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 10]
```diff
@@ -4106,10 +4105,73 @@ BOOST_AUTO_TEST_CASE(struct_delete_struct_in_mapping)
...
+			function A()
+			{
+				this.call("123");
+			}
+		}
+		contract B {
+			uint public test = 1;
+			function testIt()
+			{
+				A a;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: This does not call a constructor. You need `A a = new A();`.

File: test/libsolidity/solidityExecutionFramework.h [hunk 11]
```diff
@@ -40,142 +40,159 @@ namespace test
...
-	template <class FirstArg, class... Args>
-	static bytes encodeArgs(FirstArg const& _firstArg, Args const&... _followingArgs)
-	{
-		return encode(_firstArg) + encodeArgs(_followingArgs...);
-	}
-	static bytes encodeArgs()
-	{
-		return bytes();
-	}
+    ExecutionFramework() { g_logVerbosity = 0; }
```
Comment (chriseth) [inline test/libsolidity/solidityExecutionFramework.h]: You changed the indentation style here.

File: libsolidity/CompilerContext.h (hunk 7 above)
Comment (LianaHus) [inline libsolidity/CompilerContext.h]: hmm I'm not sure. it just returns a new created error tag. Are you sure it should be "the"?

Comment (chriseth) [inline libsolidity/CompilerContext.h]: Hm, they compare as equal but are copies of different objects, so you have a point there.

File: test/libsolidity/solidityExecutionFramework.h [hunk 12]
```diff
@@ -40,23 +40,27 @@ namespace test
 class ExecutionFramework
 {
 public:
-	ExecutionFramework() { g_logVerbosity = 0; }
+ExecutionFramework() { g_logVerbosity = 0; }
```
Comment (chriseth) [inline test/libsolidity/solidityExecutionFramework.h]: Indentation

File: test/libsolidity/solidityExecutionFramework.h [hunk 13]
```diff
@@ -40,23 +40,27 @@ namespace test
...
 
-	bytes const& compileAndRun(std::string const& _sourceCode, u256 const& _value = 0, std::string const& _contractName = "")
+	bytes const& execute(std::string const& _sourceCode, u256 const& _value = 0, std::string const& _contractName = "")
```
Comment (chriseth) [inline test/libsolidity/solidityExecutionFramework.h]: Hm, this also compiles the source code and does not just execute it.

Comment (LianaHus) [inline test/libsolidity/solidityExecutionFramework.h]: do you have a good name for this function? it does the same as compile and run only doesn't check the output

Comment (LianaHus) [inline test/libsolidity/solidityExecutionFramework.h]: _safe?

**Pull request merged**